### ISSUE TITLE
SequentialSampler getting slower as time passing by

### ISSUE THREAD

**Issue by mattivi** [pytorch/pytorch#50940]
## 🐛 Bug

I am training BERT model with a huge dataset (100M rows) and using SequentialSampler. The time spent for each step is increasing with time.
In the screenshot below you see the relationship between time in hours on x axis and global step in y axis. This relation should be linear as time passing by. 
I have noticed this issue with both Huggingface Transformer library and Simpletransformers. Replacing SequentialSampler with RandomSampler this issue disappear (but RandomSampler has other problems as in issue #50689 )

![image](https://user-images.githubusercontent.com/1651448/105484496-52973d00-5ca3-11eb-802c-422628f9e5a1.png)

## To Reproduce

1. Get a huge dataset, e.g. 100M rows
2. Start training
3. Track time over steps

## Expected behavior

I expect a linear relationship between time spent and number of steps processed, or in other words a constant training time per step during the entire training procedure. 

## Environment

PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: Tesla V100-PCIE-16GB
Nvidia driver version: 418.39
cuDNN version: /usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[conda] blas 1.0 mkl
[conda] cudatoolkit 10.1.243 h6bb024c_0
[conda] mkl 2020.2 256
[conda] mkl-service 2.3.0 py37he8ac12f_0
[conda] mkl_fft 1.2.0 py37h23d657b_0
[conda] mkl_random 1.1.1 py37h0573a6f_0
[conda] numpy 1.19.2 py37h54aff64_0
[conda] numpy-base 1.19.2 py37hfa32c7d_0
[conda] pytorch 1.7.1 py3.7_cuda10.1.243_cudnn7.6.3_0 pytorch
[conda] torchaudio 0.7.2 py37 pytorch
[conda] torchvision 0.8.2 py37_cu101 pytorch

**Comment by ejguan** [comment]
@mattivi 
Hi, I think this issue is related to python arbitrary-precision integer. As the number of elements grows, the speed would grow to O(log(n)) time for each operation.
For SequentialSampler, the range is calculated by incrementing 1 for each iteration so it will become slower and slower.
https://github.com/pytorch/pytorch/blob/e843974a6e5ecb15fe505ed3cd4e0c9cc713627a/torch/utils/data/sampler.py#L66-L68

**Comment by SsnL** [comment]
I don't think that this is a python integer problem. 

There are so many moving parts here. Are you sure that it is sequential sampler? It is just essentially a range obj. It could be also caused by how your data is stored and the dataset code, for instance.

**Comment by mattivi** [comment]
Thanks for the inputs. @ejguan and @SsnL I am using [this example script](https://github.com/huggingface/transformers/blob/master/examples/language-modeling/run_mlm.py) from Huggingface Transformers to train a Language Model from scratch. Thus, dataset is handled by the efficient Huggingface Datasets library and don't think this is the issue. 
@ejguan is there a more efficient manner to have O(n)?

**Comment by SsnL** [comment]
Similar to my reply on another thread, I believe that a minimal reproducing
example is needed before we say for sure that sequential sampler is the
problem.

**Comment by mattivi** [comment]
@SsnL I recreated the example with a minimal script

```python 
import pandas as pd
import numpy as np
import string
import random
import time
from torch.utils.data.dataloader import DataLoader
from torch.utils.data.dataset import Dataset
from torch.utils.data.sampler import RandomSampler, SequentialSampler
from datasets import load_dataset

# Create different datasets of different sizes with random words 
def id_generator(size=5, chars=string.ascii_uppercase + string.digits):
    return ''.join(random.choice(chars) for _ in range(size))

sizes = [1e4, 1e5, 1e6, 1e7] # here number of samples for each dataset
for size in sizes:
    fname = "/temp/random_"+str(int(size))+".txt"
    dataset_txt = open(fname, "w" )
    for i in range(size):
        dataset_txt.write(id_generator(size=random.randint(3,8)) + '\n')
    dataset_txt.close()

# Check overall execution time for different dataset size
overall_time = []
sizes = [1e4, 1e5, 1e6, 1e7]
for samples in sizes:
    fname = "/temp/random_"+str(int(samples))+".txt"
    dataset = load_dataset("text", data_files=fname, cache_dir="/temp/cache/")
    train_sampler = SequentialSampler(dataset["train"])
    train_dataloader = DataLoader(dataset["train"], batch_size=500, sampler=train_sampler)
    num_update_steps_per_epoch = len(train_dataloader)
    num_examples = len(train_dataloader.dataset)
    
    start_time = time.time()
    for step, inputs in enumerate(train_dataloader):
        continue
    end_time = time.time() - start_time
    
    overall_time.append(end_time)
    print("Num samples: %d \t Execution time: %3.7f"%(samples,end_time))
    
df_sequential = pd.DataFrame(sizes, overall_time).reset_index()
df_sequential.columns = ["overall_time", "sizes"]

df_sequential.plot("sizes","overall_time", logx=True, logy=True)

```
The above provide the following, which is not a direct relation.
Num samples: 10000 	 Execution time: 0.1054964
Num samples: 100000 	 Execution time: 1.0990505
Num samples: 1000000 	 Execution time: 12.0215354
Num samples: 10000000 	 Execution time: 257.7496088

![image](https://user-images.githubusercontent.com/1651448/107656433-0fb1ff00-6c7c-11eb-984e-a05c63f9ce24.png)


We can also see that as time is passing by, time per steps is increasing. Shouldn't be constant with SequentialSampler? Where else do you suggest to check? Increasing num_workers in DataLoader will speed it up, but increasing time remains.
```python 
from torch.utils.data.dataset import IterableDataset

from collections import defaultdict
overall_time = []
sizes = [1e7]
steps = defaultdict(list)
times = defaultdict(list)
for samples in sizes:
    fname = "/mapr/projects/opp/CLM/PSM/data_lm/random_"+str(int(samples))+".txt"
    dataset = load_dataset("text", data_files=fname, cache_dir="/mapr/projects/opp/CLM/PSM/data_lm/cache/")
    train_sampler = SequentialSampler(dataset["train"])
    train_dataloader = DataLoader(dataset["train"], 
                                  batch_size=500, 
                                  sampler=train_sampler, 
                                  num_workers=2
                                 )    
    start_time_overall = time.time()
    start_time = time.time()
    for step, inputs in enumerate(train_dataloader):
        if (step % 100 == True) & (step > 1):
            steps[int(samples)].append(step-1)
            end_time = time.time() - start_time
            times[int(samples)].append(end_time)
            start_time = time.time()
        pass
    end_time_overall = time.time() - start_time_overall
    overall_time.append(end_time_overall)
    print("Num samples: %d \t Execution time: %3.7f"%(samples,end_time_overall))

```
Num samples: 10000000 	 Execution time: 135.2013550

![image](https://user-images.githubusercontent.com/1651448/107670533-bd2c0f00-6c8a-11eb-831c-4575575d8625.png)

Any hint where can I do further analysis?

**Comment by SsnL** [comment]
well, your dataset could still very well matter a lot. I would say that we can blame the SequentialSampler if, e.g., the dataset is just a dummy object that reads nothing from the disk (maybe `class D: def __getitem__(self, x): return 0`) and the problem still happens.

**Comment by mattivi** [comment]
The dataset I used is a synthetic one, made of 1 word per line, word is made of random chars, as in the code above. 
If SequentialSampler is not the problem, where can it be to further investigate? Any other pytorch colleague that can help?

**Comment by matiasvergaras** [comment]
Hi,

I ran into the same problem, but with RandomSampler instead of SequentialSampler. Specifically, I was doing the following:

`
df_encoded = TensorDataset(input_ids, attention_masks, labels)
sampler = RandomSampler(df_encoded)
`

The solution was not to instantiate the random sampler with the dataframe directly, but with a sequence:

`
df_encoded = TensorDataset(input_ids, attention_masks, labels)
sampler = RandomSampler(np.arange(1, len(df_encoded)))
`

(Idea that I extracted from your other issue, since it seemed strange that it happened to you with sequentialSampler and to me with RandomSampler). 

Hope it helps.
